src/main/scala/eventstore/PersistentSubscriptionActor.scala

```diff
@@ -0,0 +1,211 @@
+  }
+
+  def reading(last: Last, next: Next, ready: Boolean): Receive = {
+    def rcv(ready: Boolean): Receive = {
+      def read(events: List[Event], n: Next, endOfStream: Boolean) = {
+        val l = process(last, events)
+        if (endOfStream) subscribing(l, n)
+        else whenReady(reading(l, n, ready = false), ready)
+      }
+
+      rcvReadCompleted(read) or
+        rcvStreamNotFound(subscribing(last, next)) or
+        rcvFailure
+    }
+
+    readEventsFrom(next)
+    rcv(ready) or rcvReady(rcv(ready = true))
+  }
+
+  def catchingUp(last: Last, next: Next, subscriptionNumber: Next, stash: Queue[Event]): Receive = {
+    def catchUp(subscriptionNumber: Next, stash: Queue[Event]): Receive = {
+      def read(events: List[Event], n: Next, endOfStream: Boolean) = {
+        if (endOfStream) liveProcessing(process(last, events), stash)
+        else {
+          @tailrec def loop(events: List[Event], last: Last): Receive = events match {
+            case Nil => catchingUp(last, n, subscriptionNumber, stash)
+            case event :: tail =>
+              val number = event.record.number
+              if (last.exists(_ >= number)) loop(tail, last)
+              else if (number > subscriptionNumber) liveProcessing(last, stash)
+              else {
+                toClient(event)
+                loop(tail, Some(number))
+              }
+          }
+          loop(events, last)
+        }
+      }
+
+      def eventAppeared(event: Event) = {
+        catchUp(subscriptionNumber, stash enqueue event)
+      }
+
+      def subscribed(number: Last) = {
+        catchUp(subscriptionNumber, Queue())
+      }
+
+      receiveEventAppeared(eventAppeared) or
+        rcvSubscribeCompleted(subscribed) or
+        rcvReadCompleted(read) or
+        rcvFailureOrUnsubscribe
+    }
+
+    readEventsFrom(next)
+    catchUp(subscriptionNumber, stash)
+  }
+
+  override def process(last: Last, event: Event): Last = {
+    val number = event.record.number
+    if (last.exists(_ >= number)) last
+    else {
+      toClient(event)
+      Some(number)
+    }
+  }
+
+  def readEventsFrom(number: Next): Unit = {
+    val read = ReadStreamEvents(
+      streamId = EventStream.Id(streamId.streamId),
+      fromNumber = number,
+      maxCount = settings.readBatchSize,
+      direction = Forward,
+      resolveLinkTos = settings.resolveLinkTos,
+      requireMaster = settings.requireMaster
+    )
+    toConnection(read)
+  }
+
+  def rcvReadCompleted(receive: (List[Event], Next, Boolean) => Receive): Receive = {
+    case ReadStreamEventsCompleted(events, n: Next, _, endOfStream, _, Forward) =>
+      if (autoAck && this.subscriptionId.nonEmpty) toConnection(Ack(this.subscriptionId.get, events.map(_.data.eventId)))
+      context become receive(events, n, endOfStream)
+  }
+
+  def rcvSubscribeCompleted(receive: Last => Receive): Receive = {
+    case PS.Connected(subId, _, eventNum) =>
+      this.subscriptionId = Some(subId)
```

Comment:
This is a good comment, but it exposed and underlying flaw in my understanding of the persistent subscription. I thought it would function similarly to the Stream Subscription, but that isn't quite correct. From the get-go, the persistent connection should subscribe to stream - otherwise there will be no subscription ID associated with the persistent subscription to ack with. Is this understanding correct? If so, I have an idea on how I could simplify the code.